### ISSUE TITLE
@babel/parser -> Flow Type Error when 

### ISSUE THREAD

**Issue by kubilaykiymaci** [babel/babel#10651]
## Bug Report

When i am getting bundle, @babel/parser is failed. (this has been since upgrading version 7.7.0)

```js
(data: any): void => onSwipeValueChange && onSwipeValueChange({ ...data, key })
```

**Environment**
- Babel version(s): [v7.6.4]
- Node/npm version: [v8.12.0]
- OS: [OSX]

**Additional context/Screenshots**

SyntaxError: /***/Documents/projects/gitrepos/***/index.js: Unexpected token (242:34)
 at Object.raise (/****/***/node_modules/@babel/parser/lib/index.js:6931:17)

**Comment by JLHwung** [comment]
I could not reproduce this error on [REPL](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=true&loose=false&code_lz=BQEwhgLmBcAEYDsCeBKOA3A9gSxLAvAHyyYIDKA7tgA4CmAamADYCutAwgBaIDmtsAMgElyVOo1YduCPsADesAHTLwUADSwA1rSSwAviiA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=flow%2Cenv&prettier=true&targets=&version=7.7.1&externalPlugins=), could you attach your babel config and complete error stack trace?

**Comment by kubilaykiymaci** [comment]
**my babel config**
```
{
  "presets": [
    "module:metro-react-native-babel-preset"
  ],
  "plugins": [
    [
      "module-resolver",
      {
        "root": [
          "./src",
          "./public"
        ]
      }
    ],
    ["@babel/plugin-proposal-decorators", { "legacy": true }]
  ]
}
```

**full of error stack**
```
error SyntaxError: 
***/swipelistview/index.js: Unexpected token (242:34)

  240 |           onSwipeValueChange={
  241 |             onSwipeValueChange
> 242 |               ? (data: any): void => onSwipeValueChange && onSwipeValueChange({ ...data, key })
      |                                   ^
  243 |               : null
  244 |           }
  245 |           ref={(row: any): void => (this._rows[key] = row)}. Run CLI with --verbose flag for more details.
SyntaxError: ***/swipelistview/index.js: Unexpected token (242:34)

  240 |           onSwipeValueChange={
  241 |             onSwipeValueChange
> 242 |               ? (data: any): void => onSwipeValueChange && onSwipeValueChange({ ...data, key })
      |                                   ^
  243 |               : null
  244 |           }
  245 |           ref={(row: any): void => (this._rows[key] = row)}
    at Object.raise 
**/reactnative/node_modules/@babel/parser/lib/index.js:6931:17)
    at Object.unexpected 
**/node_modules/@babel/parser/lib/index.js:8324:16)
    at Object.parseExprAtom 
**/reactnative/node_modules/@babel/parser/lib/index.js:9583:20)
    at Object.parseExprAtom 
**/reactnative/node_modules/@babel/parser/lib/index.js:4083:20)
    at Object.parseExprSubscripts 
**reactnative/node_modules/@babel/parser/lib/index.js:9166:23)
    at Object.parseMaybeUnary 
**/reactnative/node_modules/@babel/parser/lib/index.js:9146:21)
    at Object.parseMaybeUnary 
**/reactnative/node_modules/@babel/parser/lib/index.js:9123:28)
    at Object.parseExprOps
**/reactnative/node_modules/@babel/parser/lib/index.js:9012:23)
    at Object.parseMaybeConditional 
**/reactnative/node_modules/@babel/parser/lib/index.js:8985:23)
    at Object.parseMaybeAssign 
**/reactnative/node_modules/@babel/parser/lib/index.js:8931:21)
```

**Comment by nicolo-ribaudo** [comment]
As a workaround, wrapping that arrow function in parentheses might fix the issue.

I'm assigning this to me since I initially fixed this issue more than an year ago, and then reintroduced it :sweat_smile:

**Comment by kubilaykiymaci** [comment]
@nicolo-ribaudo  i'm looking forward to your fix, thank you very much :)

**Comment by JLHwung** [comment]
Fixed in 7.7.2.
